### ISSUE TITLE
runningAccumulate function with reset key

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#6528]
**Use case**
We have `runningAccumulate` function to perform cumulative aggregations, for example: number of distinct users from the begginning up to each day. What if we want to reset accumulations for each month - to calculate number of distinct users from the first day of month up to each day of that month?

**Describe the solution you'd like**
Add a second parameter - "reset key". If it changes, reset the accumulator.

**Comment by zombiemonkey** [comment]
This is a great idea. Can it also be added to runningDifference* functions and the new neighbour function? https://github.com/yandex/ClickHouse/pull/5925.

I'd imagine in the neighbour function the behavior would be that if the key for the neighbouring value was not the same as the current row key then it would return NULL.

**Comment by alexey-milovidov** [comment]
> I'd imagine in the neighbour function the behavior would be that if the key for the neighbouring value was not the same as the current row key then it would return NULL.

Yes, great idea!

**Comment by alex-krash** [comment]
My PR on `neighbour()` function is not merged yet, so we may apply some additional logic here.
But it is not clear for me, what behavior you expect? Could you, please, provide an example of input-output?

**Comment by filimonov** [comment]
Now:
```
SELECT 
    *, 
    runningDifference(x)
FROM 
(
    SELECT 
        number, 
        arrayJoin([1, 2, 3, 5, 8]) AS x
    FROM numbers(3) 
) 

┌─number─┬─x─┬─runningDifference(x)─┐
│      0 │ 1 │                    0 │
│      0 │ 2 │                    1 │
│      0 │ 3 │                    1 │
│      0 │ 5 │                    2 │
│      0 │ 8 │                    3 │
│      1 │ 1 │                   -7 │
│      1 │ 2 │                    1 │
│      1 │ 3 │                    1 │
│      1 │ 5 │                    2 │
│      1 │ 8 │                    3 │
│      2 │ 1 │                   -7 │
│      2 │ 2 │                    1 │
│      2 │ 3 │                    1 │
│      2 │ 5 │                    2 │
│      2 │ 8 │                    3 │
└────────┴───┴──────────────────────┘

15 rows in set. Elapsed: 0.003 sec.
```
Should be:
```
SELECT 
    number, 
    runningDifference(x,number)
FROM 
(
    SELECT 
        number, 
        arrayJoin([1, 2, 3, 5, 8]) AS x
    FROM numbers(3) 
) 

┌─number─┬─x─┬─runningDifference(x)─┐
│      0 │ 1 │                    0 │
│      0 │ 2 │                    1 │
│      0 │ 3 │                    1 │
│      0 │ 5 │                    2 │
│      0 │ 8 │                    3 │
│      1 │ 1 │                    0 │
│      1 │ 2 │                    1 │
│      1 │ 3 │                    1 │
│      1 │ 5 │                    2 │
│      1 │ 8 │                    3 │
│      2 │ 1 │                    0 │
│      2 │ 2 │                    1 │
│      2 │ 3 │                    1 │
│      2 │ 5 │                    2 │
│      2 │ 8 │                    3 │
└────────┴───┴──────────────────────┘

15 rows in set. Elapsed: 0.003 sec.
```

**Comment by alex-krash** [comment]
@filimonov , got it, we are going to implement some kind of OVER () section from window functions.

**Comment by zombiemonkey** [comment]
Have realized this isn't going to work because of processing per block. Is there any way to force consecutive blocks to synchronize?

**Comment by alexey-milovidov** [comment]
It's implemented by @kononencheg in #8326

**Comment by alexey-milovidov** [comment]
@zombiemonkey 

> Have realized this isn't going to work because of processing per block. Is there any way to force consecutive blocks to synchronize?

Yes, all "stateful" functions: `runningDifference`, `runningAccumulate`, `neighbor` are working semi-legally from architecture standpoint. It should be fixed in more fundamental way.
